feat: Emit events in the involved objects namespace

ctrought opened the pull request:
**What this PR does / why we need it**: 

Events are currently emitted in the Gatekeeper namespace. It would be beneficial to have events emitted in the namespace where the involved object is so that users maintaining those namespaces are aware of them whether it is generated during the admission or in the audit process. It might also be standard/expected that events are emitted in the same namespace as the involved object.
 
https://pkg.go.dev/k8s.io/client-go/tools/record#EventRecorder
```
// The resulting event will be created in the same namespace as the reference object.
```
When there is no namespace (e.g. cluster resources) the event is emitted in the default namespace which is expected.

**Which issue(s) this PR fixes**:
fixes #2230 

<!--
**Is the PR title following semantic convention?**
Please refer to [Semantic types](https://github.com/open-policy-agent/gatekeeper/blob/master/.github/semantic.yml) to view accepted title convention to satisfy this status check.  
-->

<!--
**Are you making changes to Gatekeeper Helm chart?**
Please refer to [Contributing to Helm Chart](https://open-policy-agent.github.io/gatekeeper/website/docs/help#contributing-to-helm-chart) for modifying the Helm chart.
-->

<!--
**Are you making changes to Gatekeeper docs?**
Please see [Contributing to Docs](https://open-policy-agent.github.io/gatekeeper/website/docs/help#contributing-to-docs) 
-->

Signed-off-by: Craig Trought <k8s@trought.ca>

maxsmythe commented:
@ritazh or @sozercan are probably best positioned to have an opinion on this one

sozercan commented:
@ritazh do you remember why we are using gatekeeper's namespace instead of involved object's?

maxsmythe commented:
I'm trying to remember this myself.

I think there may have been some notion of security concerns... e.g. it may be possible for a malicious namespace admin to delete the event, which would change the security posture of events.

Or maybe it was making it easy to see all events in one space?

ritazh commented:
I think we were following the current rbac for violations as part of an audit (constraint is cluster level resource) and as part of the Gatekeeper pod logs (Gatekeeper namespace). Both are only accessible to Gatekeeper policy and cluster admins. 

>  It would be beneficial to have events emitted in the namespace where the involved object is so that users maintaining those namespaces are aware of them whether it is generated during the admission or in the audit process.

This makes sense. Do we want to generate events in both the GK namespace (for the policy/cluster admins) and in the involved object namespace?

ctrought commented:
> This makes sense. Do we want to generate events in both the GK namespace (for the policy/cluster admins) and in the involved object namespace?

Hmmm, I think it would be redundant to have the events duplicated. For clusters with thousands or even tens of thousands of events I personally wouldn't want the convenience of viewing all from one namespace at the cost of having them duplicated them in etcd... but to make it simpler to continue to view all violations a common gatekeeper label could be added to events so policy admins or cluster-admins can list all gatekeeper events in the cluster using a label selector.

> which would change the security posture of events as a source-of-truth (not that they're great for that anyway due to throttling).

I agree with @maxsmythe that events are not really a great source of truth for known violations anyways. But if that is the only way to view all policy violations right now (provided the user enables it) and there is concern about users removing events, users could always create a GK policy to prevent them from deleting events with the proposed GK labels 😆

ritazh commented:
>  I personally wouldn't trade for the convenience of viewing all events from one namespace at the cost of having them duplicated in etcd... but to make it simpler to continue to view all violations a common gatekeeper label could be added to the events so policy admins or cluster-admins can list all gatekeeper events in the cluster using a label selector.

+1 on label selector to view all gatekeeper events. Curious how others feel about allowing users to delete the violating events in their namespace.

> But if that is the only way to view all policy violations right now 

@ctrought Have you evaluated the audit pod logs? Audit pod emits JSON-formatted audit logs to stdout https://open-policy-agent.github.io/gatekeeper/website/docs/audit#audit-logs that should have all the policy violations.

maxsmythe commented:
> Curious how others feel about allowing users to delete the violating events in their namespace.

I think it depends on the intended purpose of the events. If they are merely informational, then the only concerns about who has access where would be around preventing unintentional data leakage.

ctrought commented:
> @ctrought Have you evaluated the audit pod logs? Audit pod emits JSON-formatted audit logs to stdout https://open-policy-agent.github.io/gatekeeper/website/docs/audit#audit-logs that should have all the policy violations.

I have and I think that's great! It's a bit cumbersome to extract the current violations though (since last audit run). Need to ship these to an external sink of some kind (elasticsearch, alertmanager, etc.). This is actually one of the reasons I submitted the PR to change the namespace because if we can get the events moved to the involved objects namespace I'm planning to use an event exporter to ship these to alertmanager or some other 'sink' where I should get a compiled feed of all the violations and can send them to slack, email, or whatever receiver we'd like. Having these in the involved namespace will allow us to route the violation/alert accordingly based on source namespace.

> +1 on label selector to view all gatekeeper events. 

Sounds good, I will update the PR. 😃

ritazh commented:
@ctrought Per the 11/2/2022 community call, we think it would help to add a flag to allow operators to configure to create the events in the violating resource namespace instead of the default gatekeeper namespace.

ctrought commented:
> 

> @ctrought Per the 11/2/2022 community call, we think it would help to add a flag to allow operators to configure to create the events in the violating resource namespace instead of the default gatekeeper namespace.

I did consider it being an option via a flag, but curious what's the reason to have them located in the GK namespace? Based on API docs, to me it suggests events are meant to be created in the involved object namespace.

In OpenShift when you click on the event in the GK namespace it returns a 404 because it expects the involved object to be in the namespace where the event was generated. Are there examples where events for namespaces resources are generated in a namespace different than the involved object?

ritazh commented:
The flag should indicate all namespaces should be set to the violating resource's namespace instead of gatekeeper-system.
What about cluster scoped resources?

ctrought commented:
> @ctrought Thanks for opening this PR and discussing options. Are you still working on this PR or blocked by anything?

Hey @ritazh , I will look at this again shortly. 👍

ctrought commented:
@ritazh was there any preference on a distinct flag for each audit & webhook and helm, or distinct flag for audit and webhook with a shared option in helm that sets both?

ritazh commented:
> @ritazh was there any preference on a distinct flag for each audit & webhook and helm, or distinct flag for audit and webhook with a shared option in helm that sets both?

+1 for distinct flags for audit and webhook to make this more flexible.

ctrought commented:
I've updated the PR, in addition to the feedback made a couple other tweaks
- Added distinct flags for audit/webhook to enable event creation in the involved objects namespace
  - if there is no namespace for the object with the violation (ie. cluster resources) the event is emitted in the namespace gatekeeper is installed in
- Remove resource namespace from event message if the event is emitted in the involved namespace or it's a cluster scoped resource as it's redundant
- Use the actual involved objects ObjectReference metadata in the event
  - I'm not too sure if there was any reason why the old custom generated UID was used (`rkind + "/" + rnamespace + "/" + rname + "/" + ckind + "/" + cnamespace + "/" + cname`), perhaps it just wasn't as relevant since the events were only ever emitted in the GK namespace. The constraint information is already included in annotations in the event for additional context so the information is still available with it removed from the uid. Kubernetes will correlate events with the object when the metadata of the events involved object matches one in the cluster. For example with the change if the option to emit audit/webhook events in the involved namespace is enabled violations for a deployment will be listed under events for the object when you run `kubectl describe deploy my-deployment` which is valuable for end users

ctrought commented:
> +1 on label selector to view all gatekeeper events

Looked into this in the process, not currently possible with EventRecorder interface but there is a PR open. If it's implemented I can come back to it.

https://github.com/kubernetes/kubernetes/pull/115058

ctrought commented:
@ritazh anything you need from me to get this merged? thx

ctrought commented:
@maxsmythe @sozercan hoping this can be included in the 3.12 milestone?

ritazh commented:
Apologies on the delay. Reviewing now.

ritazh commented on pkg/audit/manager.go:
```diff
@@ -1061,19 +1064,33 @@ func emitEvent(constraint *unstructured.Unstructured,
-	ref := getViolationRef(gkNamespace, resourceGroupVersionKind.Kind, rname, rnamespace, constraint.GetKind(), constraint.GetName(), constraint.GetNamespace())
+	enamespace := gkNamespace
+	if *auditEventsInvolvedNamespace && len(rnamespace) > 0 {
+		enamespace = rnamespace
```
Perhaps add a note in the readme and docs that cluster scoped resources would create events in the GK namespace.

ritazh replied:
Can you add a unit test for emitEvent to test all the namespace logic?

ritazh commented on pkg/audit/manager.go:
```diff
@@ -1061,19 +1064,33 @@ func emitEvent(constraint *unstructured.Unstructured,
+	ref := getViolationRef(enamespace, resourceGroupVersionKind.Kind, rname, rrv, ruid)
 
-	eventRecorder.AnnotatedEventf(ref, annotations, corev1.EventTypeWarning, reason, "Timestamp: %s, Resource Namespace: %s, Constraint: %s, Message: %s", timestamp, rnamespace, constraint.GetName(), message)
+	if *auditEventsInvolvedNamespace || len(rnamespace) == 0 {
```
should `auditEventsInvolvedNamespace` be false here?

ctrought replied:
same as below, it's just to avoid printing the resources namespace if the events are being generated in the same namespace (redundant as we already know) or there is no namespace (cluster scoped resources).

ritazh replied:
For the case where the feature is not enabled, we should keep it the same as before to reduce impact for existing users. Can you restore it back to what it was?

ctrought replied:
done, behaviour is reverted when flag disabled

ritazh commented on pkg/audit/manager.go:
```diff
@@ -1061,19 +1064,33 @@ func emitEvent(constraint *unstructured.Unstructured,
-		Namespace: gkNamespace,
+		Namespace: enamespace,
+	}
+	if len(ruid) > 0 && len(rrv) > 0 {
```
Is there a case where you only want to set one of them if only one of them is valid?

ctrought replied:
tbh I'm not sure, I thought about it initially but I couldn't think why a resource version would not exist if the object has been assigned a UID.

ritazh commented on pkg/audit/manager_test.go:
```diff
@@ -199,43 +198,41 @@ func Test_getViolationRef(t *testing.T) {
+				Kind:      		 "Service",
+				Name:     		 "my-service",
+				Namespace: 		 "kube-system",
+				ResourceVersion: "123456",
```
Maybe add a test case for `len(ruid) > 0 && len(rrv) > 0 `?

ritazh commented on pkg/webhook/policy.go:
```diff
@@ -290,24 +293,19 @@ func (h *validationHandler) getValidationMessages(res []*rtypes.Result, req *adm
+
+			ref := getViolationRef(enamespace, req.AdmissionRequest.Kind.Kind, resourceName, obj.GetResourceVersion(), obj.GetUID())
+
+			if *admissionEventsInvolvedNamespace || len(req.AdmissionRequest.Namespace) == 0 {
```
Same here, should `admissionEventsInvolvedNamespace` be false here?

ctrought replied:
This just changes the way the event message is output slightly. It checks if the events are generated in the involved objects namespace (flag true) or this is a cluster scoped resource (no namespace) then we do not need to output the resources namespace in the events message... imo this is because

1. When the flag is enabled and we view an event, the event is already generated in the resources namespace so it's redundant. 
2. Cluster scoped events don't have a namespace

ritazh commented on pkg/webhook/policy.go:
```diff
@@ -624,13 +622,17 @@ func createReviewForResultant(obj *unstructured.Unstructured, ns *corev1.Namespa
+	ref := &corev1.ObjectReference{
 		Kind:      rkind,
 		Name:      rname,
-		UID:       types.UID(rkind + "/" + rnamespace + "/" + rname + "/" + ckind + "/" + cnamespace + "/" + cname),
```
why do we want to remove constraint information here? I think that is useful to know.

ctrought replied:
This information is still all contained in the events annotations so it is not lost :)
```golang
	annotations := map[string]string{
		"process":                    "audit",
		"auditTimestamp":             timestamp,
		logging.EventType:            "violation_audited",
		logging.ConstraintGroup:      constraint.GroupVersionKind().Group,
		logging.ConstraintAPIVersion: constraint.GroupVersionKind().Version,
		logging.ConstraintKind:       constraint.GetKind(),
		logging.ConstraintName:       constraint.GetName(),
		logging.ConstraintNamespace:  constraint.GetNamespace(),
		logging.ConstraintAction:     string(enforcementAction),
		logging.ResourceGroup:        resourceGroupVersionKind.Group,
		logging.ResourceAPIVersion:   resourceGroupVersionKind.Version,
		logging.ResourceKind:         resourceGroupVersionKind.Kind,
		logging.ResourceNamespace:    rnamespace,
		logging.ResourceName:         rname,
	}
```

The main reason to use the involved objects real UID would be so these events can be discovered by existing tools that link events to the resources they are about. For example the describer process in kubectl searches events in the namespace for matching `name` `kind` and `uid` and will display the events alongside the resource which is useful information for end users and brings greater visibility to possible issues. Below is what that might look like.

```sh
$ kubectl describe deployment grafana-operator-controller-manager
Name:               grafana-operator-controller-manager
Namespace:          offending-namespace
...
...
...
Events:
  Type     Reason          Age                From              Message
  ----     ------          ----               ----              -------
  Warning  AuditViolation  9m (x28 over 27h)  gatekeeper-audit  Constraint: must-have-probes, Message: [Implied by expand-deployments] Container <kube-rbac-proxy> in this <Pod> has no <readinessProbe> and is selected by service <grafana-operator-controller-manager-metrics-service> with targetPort(s) ["https"].
```

This of course only works if the events are generated in the involved objects namespace. If the arbitrary (current) UID remains events will not visible on objects when using kubectl or other tools that associate resources with events that are about them.

https://github.com/kubernetes/kubectl/blob/master/pkg/describe/describe.go#L280
https://github.com/kubernetes/kubectl/blob/master/pkg/describe/describe.go#L5444

ritazh replied:
Thanks for sharing discoverability by existing tools. That makes sense. 

The UID is concatenated as such today is due to the concerns described [here](https://github.com/open-policy-agent/gatekeeper/pull/739#discussion_r459721241) for clobbering. Could the same concern be applicable here if there are multiple events generated for the same violating resource?

Similar to an earlier comment, for the case where the flag is not enabled, lets restore it back to what it was to reduce impact especially because we know clobbering could happen.

ritazh commented on website/docs/customize-startup.md:
```diff
@@ -23,11 +23,15 @@ The `--disable-opa-builtin` flag disables specific [OPA built-ins functions](htt
-There are three types of events that are emitted by Gatekeeper when the above flags are enabled:
+The `--admission-events-involved-namespace` flag controls which namespace admission events will be created in. When set to `true` admission events will be created in the involved objects namespace violating the constraint, if the object has no namespace (ie. cluster scoped resources) they will be created in the namespace gatekeeper is installed in. Setting to `false` will cause all admission even ...
+
+The `--audit-events-involved-namespace` flag controls which namespace audit events will be created in. When set to `true` audit events will be created in the involved objects namespace violating the constraint, if the object has no namespace (ie. cluster scoped resources) they will be created in the namespace gatekeeper is installed in. Setting to `false` will cause all audit events to be created ...
```
As noted above, pls add something for cluster-scope resources.

ctrought replied:
> `if the object has no namespace (ie. cluster scoped resources) they will be created in the namespace gatekeeper is installed in. Setting to "false" will cause all audit events to be created in the gatekeeper namespace.`

It's there already, would you like it amended?

sozercan commented:
@ctrought looks like tests are failing, can you ptal?

sozercan commented:
@ctrought ptal the failing tests. we are aiming to release rc on 3/24.

ritazh commented:
@ctrought _bump_ so we can get this PR reviewed and merged for the upcoming release

eshaanm25 commented:
The unit tests pass on the [fork](https://github.com/ctrought/gatekeeper/actions/runs/4497140536/jobs/7912457419) and a [local machine](https://user-images.githubusercontent.com/37758843/227129089-b520bea3-18c7-453f-a072-12c1395f1614.png). The job may need to be re-run for this PR.

ritazh reviewed (approved):
LGTM

maxsmythe reviewed (approved):
LGTM
